File: jpp/lab5/program.hs
fact 0 = 1
fact 1 = 1
fact n = n * fact (n - 1)

myGcd a b = if a < b then myGcd' b a else myGcd' a b

myGcd' a 0 = a
myGcd' a b = myGcd' b (a `mod` b)

diof a b c = if c `mod` d == 0 then (x * (c `div` d), y * (c `div` d)) else error "No solution"
  where
    (x, y) = diof' a b
    d = myGcd a b

diof' a b = if b == 0 then (1, 0) else (y, x - (a `div` b) * y)
  where
    (x, y) = diof' b (a `mod` b)

File: jpp/lab5/program.pl
factorial(0, 1) :- !.
factorial(N, X) :-
  N > 0,
  N1 is N - 1,
  factorial(N1, X1),
  X is N * X1.

gcd(X, 0, X) :- !.
gcd(X, Y, G) :-
  Y > 0,
  Y1 is X mod Y,
  gcd(Y, Y1, G).

diof(A, B, C, M, N) :-
  gcd(A, B, D),
  gcd(C, D, D),
  diof1(A, B, X, Y),
  M is X * (C // D),
  N is Y * (C // D).

diof1(_, 0, 1, 0) :- !.
diof1(A, B, X, Y) :-
  C is A mod B,
  diof1(B, C, X1, Y1),
  X is Y1,
  Y is X1 - (A // B) * Y1.
